wizard.py

```diff
@@ -0,0 +1,1037 @@
+    # Add posted comment to ssc_data
+    ssc_data["comment"] = comment
+
+    # Persist and redirect to next page
+    _persist_session_subdocument_ts({"ssc": ssc_data})
+    return redirect(url_for("functionaries"))
+
+
+
+  # Query all session data (posted on vcs, building, qa, ... pages)
+  session_data = _get_session_document()
+
+  # Query any existing software supply chain data (posted on this page)
+  ssc_data = session_data.get("ssc", {})
+  ssc_last_modified = ssc_data.get("last_modified", 0)
+
+  # Assume we don't have to show the refresh dialog (explained below)
+  show_refresh_dialog = False
+
+  # Only (re)generate software supply chain graph from data the user has posted
+  # on previous pages, if there is no ssc data from this page in the db or the
+  # user has sent he `refresh` parameter
+  if not ssc_data or request.args.get("refresh"):
+    ssc_data = session_to_ssc(session_data)
+
+  # Otherwise we serve existing ssc data
+  else:
+    # If existing ssc data is older than any of stored vcs/building/qa/package
+    # data we still serve the stored ssc data but additionally show a
+    # "Do you want to re-generate the software supply chain?" dialog with a
+    # link that includes the `refresh` get parameter
+    for subdocument in ["vcs", "building", "qa", "package"]:
+      data_last_modified = session_data.get(subdocument,
+          {}).get("last_modified", 0)
+
+      if ssc_last_modified < data_last_modified:
+        show_refresh_dialog = True
+        break
+
+  return render_template("software_supply_chain.html",
+      ssc_data=ssc_data, show_refresh=show_refresh_dialog)
+
+
+@app.route("/functionaries", methods=["GET", "POST"])
+@with_session_id
+def functionaries():
+  """Step 6.
+  On get serve functionary keys upload and keys dropzone.
+  On post store comment and redirect to next page.
+
+  The actual functionary public key upload uses a different view queried
+  with ajax (ajax_upload_key).
+
+  """
+  functionaries = _get_session_subdocument("functionaries")
+  if request.method == "POST":
+    functionaries["comment"] = request.form.get("comment", "")
+    _persist_session_subdocument({"functionaries": functionaries})
+
+    flash("Now tell us who is authorized to do what...",
+        "alert-success")
+    return redirect(url_for("authorizing"))
+
+  return render_template("functionaries.html", functionaries=functionaries)
+
+
+@app.route("/functionaries/upload", methods=["POST"])
+@with_session_id
+def ajax_upload_key():
+  """Ajax upload a functionary key. Key files are stored to the db in their
+  dictionary representation. """
+  functionary_key = request.files.get("functionary_key", None)
+  functionary_name = request.form.get("functionary_name", None)
+
+  if not functionary_name:
+    flash("Something went wrong: We don't know which functionary,"
+              " this key belongs to", "alert-danger")
+
+    return jsonify({"error": True})
```

Comment:
Do you want to return more specific info about the error?

Reply:
A better error message than *"Something went wrong: We don't know which functionary,  this key belongs to"* or a specific error code? Do you have something in mind? Btw. this error should only occur if the user tampers with the form or the JS.